jbpm-flow/src/main/java/org/jbpm/process/instance/impl/FeelReturnValueEvaluator.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.process.instance.impl;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jbpm.process.core.context.variable.VariableScope;
+import org.jbpm.process.instance.context.variable.VariableScopeInstance;
+import org.jbpm.workflow.instance.WorkflowProcessInstance;
+import org.kie.api.runtime.Globals;
+import org.kie.api.runtime.process.ProcessContext;
+import org.kie.dmn.feel.FEEL;
+
+public class FeelReturnValueEvaluator implements ReturnValueEvaluator, Externalizable {
+    
+    private static final long   serialVersionUID = 630l;
+
+    private String expr;
+
+    public FeelReturnValueEvaluator() {
+    }
+
+    public FeelReturnValueEvaluator(String expr) {
+        this.expr = expr;
+    }
+
+    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+        expr = in.readUTF();
+    }
+
+    public void writeExternal(ObjectOutput out) throws IOException {
+        out.writeUTF( expr );
+    }
+
+    public Object evaluate(ProcessContext context) throws Exception {
+        Map<String, Object> variables = new HashMap<>();
+        variables.put("kcontext", context);
+        
+        // insert globals into context
+        Globals globals = context.getKieRuntime().getGlobals();
+        
+        if (globals != null && globals.getGlobalKeys() != null) {
+            for (String gKey : globals.getGlobalKeys()) {
+                variables.put(gKey, globals.get(gKey));
+            }
+        }
+        if (context.getProcessInstance() != null && context.getProcessInstance().getProcess() != null) {
+            // insert process variables
+            VariableScopeInstance variableScope = (VariableScopeInstance) ((WorkflowProcessInstance)context.getProcessInstance())
+                    .getContextInstance(VariableScope.VARIABLE_SCOPE);
+    
+            if (variables != null ) {
+                variables.putAll(variableScope.getVariables());
+            }
+        }
+        FEEL feel = FEEL.newInstance();
```

Comment:
Ideally, we need to cache, because there are configurations (profiles) that might need to be enabled at instantiation. @tarilabs can you show maciej how to make sure this is using the same configuration as the DMN engine? My concern is: if we have a process invoking DMN decision services, the FEEL instance must use the same configuration for both.

Reply:
It is not a heavyweight but at least on DMN side @etirelli requires me to cache it. If it's feasible without having to rearchitecture, I would cache it.

Reply:
We usually create the FEEL instance supplying the appropriate classloader and profiles this way:
https://github.com/kiegroup/drools/blob/1ab2c0444f3bc9ea07f2dee5a1d93ab7a85249d1/kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/DMNFEELHelper.java#L58

Normally the classloader is the Kie project (kjar) rootclassloader. This is necessary to make sure FEEL can locate some external class correcty, IFF necessary.

About _profiles_ these are governed by the `kmodule.xml` or System-wise-property, by some specific properties.

Normally 1 default profile is always applied, unless the user specified a "strict' property:
https://github.com/kiegroup/drools/blob/1ab2c0444f3bc9ea07f2dee5a1d93ab7a85249d1/kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/assembler/DMNAssemblerService.java#L191

The other Profiles are taken from the configuration (`kmodule.xml` or system-wise-property)
https://github.com/kiegroup/drools/blob/1ab2c0444f3bc9ea07f2dee5a1d93ab7a85249d1/kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/assembler/DMNAssemblerService.java#L172

Having the correct set of Profiles is important, as the user might have in there their own FEEL "library" functions

Reply:
If any questions about the above don't hesitate to let me know ;)